HPCC-32452 Time security manager auth calls

timothyklemm opened the pull request:
Create an ISecManager decorator that can add internal timing spans for any security manager interface call.

Replace direct manager authorization calls with calls through the decorator. Disregard CLdapSecManager-specific calls for now.

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

rpastrana commented on system/security/shared/secmanagertracedecorator.hpp:
```diff
@@ -0,0 +1,364 @@
+    } \
+    catch (...) \
+    { \
+        Owned<IException> e(makeStringException(-1, "unknown exception")); \
```
Is the span status set to fail by the exception?

rpastrana replied:
I think this is fine.

rpastrana commented on system/security/shared/secmanagertracedecorator.hpp:
```diff
@@ -0,0 +1,364 @@
+    {
+        return decorated->unsubscribe(events, secureContext);
+    }
+    virtual bool authorize(ISecUser & user, ISecResourceList * resources, IEspSecureContext* secureContext = nullptr) override
```
what's the pattern for customizing the decoration behavior?
Let's say the esp knows it doesn't want to trigger a span for a specific secmanager call. Would that require a second TSecManagerTraceDecorator with a different method def?

timothyklemm replied:
This doesn't sound like a decorator question. As long as the ESP is making the decision to instrument, the question is how does the ESP decide which managers do or do not require instrumentation.

If we assume LDAP will always require instrumentation, we could define a new plugin configuration value (likely in the SecurityManager element wrapping the manager's configuration). This value could be passed into the decorator constructor, allowing the decorator to do what it needs to do. We could create a threaded active decorator at manager construction, or use a "null decorator" when instrumentation is not needed.

A second option could be to make the determination a function of the manager itself. I'm not sure there's an existing API to make this decision, but one could be added.

rpastrana replied:
Let me ask it a different way.
How would we suppress the spans added by the decorator if a very simple secmanager is used which doesn't need to trace its calls.

timothyklemm replied:
The ESP doesn't currently know if it is a very simple manager. There is an enumerated value that managers may return, which we could use to decide, but there are no values defined for managers outside the platform and could be unreliable. There is a label associated with the enumeration, but there are no naming conventions in place through which we could extract anything like this.

The only way the ESP will know that it is a simple manager is to ask it, or look for a new configuration value.  I've intentionally omitted the third option which is dynamically casting.

rpastrana replied:
@timothyklemm I didn't ask how the ESP can determine if a given call should be traced or not, but I think I can infer that we're stuck w/ the spans declared by the decorator for all sec manager.Thanks.

rpastrana commented on system/security/shared/secmanagertracedecorator.hpp:
```diff
@@ -0,0 +1,364 @@
+ * - If nothing else changes, create a subclasses of TSecManagerTraceDecorator, with template
+ *   parameters CLdapSecManager and ISecManager, to decorate the LDAP-specific interfaces.
+ */
+template <typename secmgr_t = ISecManager, typename secmgr_interface_t = ISecManager>
```
what would be the advantage/disadvantage of this approach vs a base implementation which all sec managers could extend?

timothyklemm replied:
Without the template, the base class holds a reference to an ISecManager instance, and any subclass holds a second reference to the same object but typed as the manager implementation class. Very minor, but I'm not a fan of holding the same value twice. The first template parameter allows the base class to hold the type needed by the subclass.

The two template parameters could be consolidated into one, reducing the complexity, if either:
- The ISecManager and CLdapSecManager interfaces are standardized.
- A new extension of ISecManager declaring the LDAP extensions is used.

I'm not married to either approach.

rpastrana reviewed (commented):
@timothyklemm good looking code. left you a couple of minor comments and some high level questions.

rpastrana commented:
> hould I decorate every manager method,

@timothyklemm I don't think decorating every method would be helpful.
In fact as we've discussed, for some managers, the 2 methods decorated might result in too many spans...

kenrowland commented on system/security/shared/secmanagertracedecorator.hpp:
```diff
@@ -0,0 +1,364 @@
+    }
+    virtual bool authorizeEx(SecResourceType rtype, ISecUser & user, ISecResourceList * resources, IEspSecureContext* secureContext = nullptr) override
+    {
+        START_SEC_MANAGER_TRACE_BLOCK("security.authorize_ex");
```
Should this be "security.authorize_ex_list" to distinguish between a single authorization and a list? It would seem that ia a long list was passed that it could skew results if it's measured the same as a single authorization. There is at least one other case where a method is overloaded to accept both a single instance and a list.

timothyklemm replied:
It can be. I thought about it because it's already a distinction made in the feature flags.

kenrowland commented on system/security/shared/secmanagertracedecorator.hpp:
```diff
@@ -0,0 +1,364 @@
+/**
+ * @brief Templated utility class to decorate an object with additional functionality.
+ *
+ * A decorator implements the same named interface as the objects it decorates. This class extends
```
trivial: objects should be object

kenrowland commented on system/security/shared/secmanagertracedecorator.hpp:
```diff
@@ -0,0 +1,364 @@
+ * interface. By implementing the same interface, a decorator is interchangeable with the object
+ * it decoratos.
+ *
+ * In less than ideal situations, the decorated object's interface is an extension of a named
```
Trivial, should "an extension" be "a subclass"

timothyklemm replied:
I think they're synonyms, but I'll change it.

kenrowland commented on system/security/shared/secmanagertracedecorator.hpp:
```diff
@@ -0,0 +1,364 @@
+ * it decoratos.
+ *
+ * In less than ideal situations, the decorated object's interface is an extension of a named
+ * interface. The decorator extends the extends the named interface, with subclasses required to
```
trivial: "extends the extends" should probably be changed a bit.

kenrowland reviewed (changes requested):
Generally looks like what we discussed. Just a few comments.

timothyklemm commented on common/workunit/workunit.cpp:
```diff
@@ -113,7 +114,7 @@ static bool checkWuScopeSecAccess(const char *wuscope, ISecManager *secmgr, ISec
     if (!secmgr || !secuser)
         return true;
-    bool ret = secmgr->authorizeEx(RT_WORKUNIT_SCOPE, *secuser, wuscope)>=required;
+    bool ret = CSecManagerTraceDecorator(*secmgr).authorizeEx(RT_WORKUNIT_SCOPE, *secuser, wuscope)>=required;
```
It may be a best practice to use a single security manager configuration (implying a single implementation) within an ESP process. ESPs can be configured to use multiple managers and configurations.

A static interface would require either passing the manager to each decorated call, negating the benefit of extending ISecManager, or a threaded security manager (similar to the threaded active span). The latter option would likely complicate the implementation by requiring manager checks in every decorated method.

I'm not sure which is preferred. Also, one of @rpastrana's comments might indirectly justify a threaded active decorator, so only one decorator would be created per service.

timothyklemm commented on esp/bindings/http/platform/httpbinding.cpp:
```diff
@@ -972,7 +973,7 @@ bool EspHttpBinding::basicAuth(IEspContext* ctx)
     }
 
-    bool authenticated = m_secmgr->authorize(*user, rlist, ctx->querySecureContext());
+    bool authenticated = CSecManagerTraceDecorator(*m_secmgr).authorize(*user, rlist, ctx->querySecureContext());
```
The decorator always adds, and releases, its own reference to the object, without affecting ownership of the reference passed to it. I believe it is safe.

As it's being used, linking might not be needed. If the decorator ever has a longer life than one manager call, it will be.

kenrowland replied:
Not being an expert with Owned and Linked, but is the reference count for the object shared between the original Owned and the Linked pointers?

kenrowland commented:
@timothyklemm There are no plans at this time to make the LDAP specific interface part of ISecManager. If having them the same significantly reduces the decorator solution and the need exists to decorate and trace LDAP security manager calls, I can investigate further. Consolidating them would be a good interim step to eventual replacement.

Commit:
HPCC-32452 Add conditional tracing for ESP security manager authorization calls.

- Define a new, common, trace option for security manager tracing. Both ESP and
  Roxie use security managers, and the flag can be shared by both.
- Refactor CEspHttpServer to create server spans before initial authorization
  requests.
- Create a security manager decorator encapsulating the logic of what to trace
  and when.
- Replace security manager authorization calls with decorated calls. This change
  excludes calls specific to the LDAP security manager.

Signed-off-by: Tim Klemm <tim.klemm@lexisnexisrisk.com>

Commit:
HPCC-32465 Add ESP support for trace level

- Define an ESP process configuration node that supports specification of global
  TraceFlags values for each ESP.
- Reserve traceLevel to request a specific trace level (0: none, 1: standard,
  2: detailed, 3: max). This replaces acting on the last observed occurrence
  of either traceNone, traceStandard, traceDetailed, and traceMax.
- Override default flag settings when not configured and a debug build.

Signed-off-by: Tim Klemm <tim.klemm@lexisnexisrisk.com>

timothyklemm commented:
@rpastrana and @kenrowland, this is still a work in progress. There are two commits, each representing a separate Jira. The first commit sets up the ESP for using trace flags. The second has the decorator using trace flags to control span creation. Do these address prior concerns?

rpastrana commented on esp/platform/espp.cpp:
```diff
@@ -577,6 +649,7 @@ int init_main(int argc, const char* argv[])
             config->bindServer(*server.get(), *server.get());
             config->checkESPCache(*server.get());
 
+            initializeTrace(config);
```
I like the symmetry w/ the initializeMetrics call, but this name can be confused with Tracing.
InitializeTraceLevelSettings ?

rpastrana reviewed (commented):
@timothyklemm a few comments on the traceflag commit

asselitx reviewed (commented):
Overall this looks like a good solution to getting the trace configuration into the ESP and for timing the secmgr calls.

timothyklemm commented on esp/platform/espp.cpp:
```diff
@@ -354,6 +358,74 @@ static void usage()
+    {
+        VStringBuffer attrName("@%s", o.name);
+        const char* value = nullptr;
+        if (!(value = ptree->queryProp(attrName)) && !(value = ptree->queryProp(attrName.setf("_%s", o.name))))
```
The underscore prefix is an alternate xpath used in loadTraceFlags. I don't know when, or if, it's used but retained it to minimize functional differences between the two functions.

Commit:
Review comments

Introduce a persistent decorator instance to the HTTP binding, ESP context, and
security handler to avoid repeated construction and destruction.

Improve the binding's handling of security manager creation failures to avoid
null dereferences when creating the decorator, or resource auth maps.

ghalliday reviewed (changes requested):
This does not seem to be the correct approach to me.  It is more complicated that I would expect for the problem that I think it is trying to solve.

The jira could do with a discussion of the potential designs, exactly what it would be beneficial to trace and why.  Adding the code in this place means that all security managers - including fixed username, or simple table of users would be tracing all authentication calls.

Similarly for ldap security managers it will be creating spans when the value is returned immediately from the cache.  We do not want to generate lots of short-lived spans, especially if they do not involve a server.

What is potentially important is tracing the requests to the LDAP server.  I imagine that can be best achieved by adding code to create a client span within the ldap code itself.  That also allows ldap-specific details to be added to the span (if there are any).

timothyklemm commented:
@ghalliday the code currently included in the PR is more complex than it should be. If https://github.com/hpcc-systems/HPCC-Platform/pull/19292 is accepted, we will be able to decorate managers as they are constructed, and no additional awareness of the decorator is required.

@rpastrana and I discussed whether it would be preferable for the ESP or the manager to be responsible for instrumentation. With the potential for 3rd party managers (not just mine, but an OAuth 2 manager is being developed), we concluded it would be better for the ESP to produce consistent output instead of hoping managers will do it for us.

To reduce the number of unnecessary spans, the feature must be explicitly turned on and only "implemented" methods, as identified by the manager, will be instrumented. Unfortunately, there doesn't seem to be a reliable way, given the current interface, to identify which managers warrant instrumentation. Also, I can't predict whether something will be cached before creating a span.

It could be possible to replace spans with timing attributes. This could be start and end times (similar to current TxSummary content), start time and duration, or just duration. Inclusion of the attributes could be conditional on a minimum duration. Retroactive span creation given a start and duration, if possible, could also be conditional on a minimum duration.

Both options would prevent grouping any spans created by the manager itself. if I was actively working on my plugin, I would probably have it creating child spans for each of its multiple HTTP and MySQL requests. Any authorize request to one of my managers could require at least 3 HTTP requests and 2 MySQL requests to complete.

rpastrana commented:
@timothyklemm as I recall there were several long conversations regarding this feature. I don't recall what might have convinced me to agree to the current proposal, but in general I'd expect the ESP to span potentially lengthy/complex calls. I'd also expect the call to have the ability to declare its own child span to represent some interesting subtask (I think you pointed out why this might be difficult to achieve, but I don't recall). Another important aspect we've emphasized is to avoid affecting existing functionality for the sake of tracing.
